src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,292 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <uv.h>
+
+#include "nvim/api/private/helpers.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/lua/executor.h"
+#include "nvim/os/keyboard.h"
+#include "nvim/vim.h"
+#include "nvim/version.h"
+
+#if defined(__linux__)
+# define XLIB_NAME "libX11.so.6"
+
+typedef void *_XDisplay;
+
+/// Holds the data of the present state of Keyboard
+typedef struct {
+  char data[18];
+} _XkbStateRec;
+typedef int _XStatus;
+
+/// The offset of the keyboard lock modes for X11
+enum { kLockedModsOffset = 9 };
+enum { _XkbUseCoreKbd = 0x0100 };
+
+/// Masks used in libX11 for different Lock states
+typedef enum {
+    kXNumLock = 0x10,  ///< Mask for NumLock
+    kXCapsLock = 0x02,  ///< Mask for CapsLock
+    kXScrollLock = 0x80,  ///< Mask for ScrollLock
+} XlibKbdLocks;
+#elif defined(WIN32)
+# include <Winuser.h>
+#elif defined(__APPLE__)
+# define Boolean Boolean_I_Dont_Care
+# include <IOKit/IOKitLib.h>
+# include <IOKit/IOReturn.h>
+# include <IOKit/hidsystem/IOHIDLib.h>
+# include <IOKit/hidsystem/IOHIDParameter.h>
+# include <CoreFoundation/CoreFoundation.h>
+# undef Boolean
+#endif
+
+/// Masks for the specific keyboard lock status
+typedef enum {
+  kNumLock = 0x01,  ///< Mask used to check if the NumLock is active
+  kCapsLock = 0x02,  ///< Mask used to check if the CapsLock is active
+  kScrollLock = 0x04,  ///< Mask used to check if the ScrollLock is active
+} KbdLocks;
+
+/// Used as a bitmask for the status of the different Lock Status
+typedef int ModMask;
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/keyboard.c.generated.h"
+#endif
+
+#if defined(__linux__)
+static bool xlib_opened = false;
+static uv_lib_t xlib;
+static _XDisplay *(*_XOpenDisplay)(void *);
+static int (*_XCloseDisplay)(_XDisplay *);
+static _XStatus (*_XkbGetState)(_XDisplay *, unsigned, _XkbStateRec *);
+
+/// A helper function to be used to load a function of libX11
```

Comment:
Documentation string looks strange and it is missing documentation about arguments.

```C
/// Load symbol from previously opened X11 library
///
/// Requires os_xlib_init() to be called successfully before.
///
/// @param[in]  name  Symbol to load. Assumes it being a function name.
/// @param[out]  fun  Location where to put load results.
/// @param[out]  err  Location where to save error.
///
/// @return true if loading was successfull, false otherwise.
```

Reply:
OK
> /// @param[in]  name  Symbol to load. Assumes it being a function name.
Shouldn't there be obtain instead of load, as you told of a reason earlier, because of the documentation of the uv_dlsym

Reply:
There was an update as second paragraph was incorrect.

Reply:
Obtain is also good.

Reply:
BTW, cite code with `@…` as code. You can invite users like this, though I think `@param` as an organization does not mind, maybe it is even actually a dummy created by GH team.